ESC deadzone

OrionOth opened the pull request:
~~Looking at BlueRobotics' data table for their T200 thruster, there are actually only so many values under 5 amps. Ciaran already limited the values that could possibly be sent to the ESCs with the servokit library, but there was another issue: the -1,1 float range from /drive/motors was mapped directly to the 0-3000 value range of the ESC, 80% of which we couldn't use at all due to the power draw being more than 5 amps. So, we just pegged the values for 40% of our input range at 1220, and another 40% of our input range at 1780. In other words, only 20% of our range was controllable. This commit improves the accuracy by changing the linear interpolation parameters such that we can actually use most of the values.~~

Furthermore, I've also implemented a deadzone on the hardware side, as according to BlueRobotics, 
> all values **below 300 RPM** are dropped to zero, as **rotation below this value is not consistent** and results in jerky motion and non-constant thrust, power draw, etc.

This is the source of the weird stopping/starting spinning on the thrusters when we have it out of the water.

OrionOth committed:
Make ESC Lerp more meaningful + ESC deadzone

Looking at BlueRobotics' data table for their T200 thruster, there are
actually only so many values under 5 amps. Ciaran already limited the
values that could possibly be sent to the ESCs with the servokit
library, but there was another issue: the -1,1 float range from
/drive/motors was mapped directly to the 0-3000 value range of the ESC,
80% of which we couldn't use at all due to the power draw being more
than 5 amps. So, we just pegged the values for 40% of our input range at
1220, and another 40% of our input range at 1780. In other words, only
20% of our range was controllable. This commit improves the accuracy by
changing the linear interpolation parameters such that we can actually
use most of the values.

OrionOth commented:
It might perhaps be more meaningful if, rather than a deadzone, the lerp consisted of two separate lerps causing values below 300 rpm to be outside of the lerp.

cTurtle98 commented:
you misunderstand what is going on

first `-1, 1`(float) is remapped into `0, 3000`(int) by my lerp function and then servokit remaps that into `1220, 1780`(int)
its linear interpolated twice. 
its not clamped
we are not loosing resolution

cTurtle98 commented:
the way the servokit library works is

`servokit.servo[thruster].angle` expects the angle you want the servo horn to point on the servo

`servokit.servo[thruster].actuation_range` sets the maximum angle the servo can go to

`servokit.servo[thruster].set_pulse_width_range` sets the pwm frequency for zero degrees and the max angle degrees


so in my code:
`servokit.servo[thruster].angle` accepts 0 degrees to 3000 degrees
where 
0 degrees is pwm frequency 1220 and 
3000 degrees is pwm frequency 1780

OrionOth commented:
okay, i think i understand. In that case, i'll make this PR just the deadzone

cTurtle98 reviewed (changes requested):
the servokit code doesnt need changning

cTurtle98 commented on src/seahawk_rov/seahawk_rov/helper_functions.py:
```diff
@@ -24,15 +24,12 @@
 
 # linear interpolation helper function
-def float_to_pwm(old_value:float)->int:
+def float_to_pwm(old_value:float, new_min:int=0, new_max:int=3000)->int:
```
why?

DerVerrater commented:
> It might perhaps be more meaningful if, rather than a deadzone, the lerp consisted of two separate lerps causing values below 300 rpm to be outside of the lerp. 

If you cut out parts of the range, it's not a linear function. It is, by definition, not a lerp.

Which is fine, just call it something else.

OrionOth committed:
Remove extra lerp

This commit makes it so the only changes in the branch is the addition
of the anti-stutter deadzone.

cTurtle98 commented:
i dont think this should be merged in

OrionOth commented:
Why do you think that? I suppose it should be put on hold until its effectiveness is proven. Tomorrow I will go and test the ROV with an oscilloscope and see if it actually matters.

DerVerrater commented on src/seahawk_rov/seahawk_rov/helper_functions.py:
```diff
@@ -56,3 +56,10 @@ def clamp(num, minimum, maximum):
+
+# deadzone helper function
+# if within range set value to something; else do nothing
+def deadzone(num, min, max, value=0):
```
Names `num` and `value` aren't the best. It's not clear which one is being deadzoned.

The code looks like it will select the `value` when the `num` is out of bounds. It should default to 0 with no option to change. This function should operate in normalized space, not the PWM frequency range. There's another converter for that step.

DerVerrater commented on src/seahawk_rov/seahawk_rov/thrust_box_servo.py:
```diff
@@ -55,4 +55,4 @@ def __init__(self, node, i2c):
     def receive_thruster(self, message:Float32MultiArray):
         for thruster in self.thruster_map:
-            self.kit.servo[thruster].angle = int(seahawk_rov.float_to_pwm(seahawk_rov.clamp(message.data[thruster], -1.0, 1.0)))
+            self.kit.servo[thruster].angle = int(seahawk_rov.deadzone(seahawk_rov.float_to_pwm(seahawk_rov.clamp(message.data[thruster], -1.0, 1.0)), 1464, 1536, 1500))
```
I think this is a good time to introduce decorators. You can use `@functionname` just ahead of a `def` statement on the previous line to wrap the function in another function.

We could have a deadzone function that has the same signature as the lerp function, and then decorate the lerp with the deadzone call. Then they'll all happen together when we call the `float_to_pwm`. It's logically identical to nesting function calls like this, but much easier to read.

DerVerrater reviewed (commented):
Branch renaming is free. This should be named for the deadzone feature it introduces.

OrionOth commented:
This isn't happening, what with continuous servo reworks and other stuff.
